pallets/asset-manager/src/migrations.rs

```diff
@@ -0,0 +1,416 @@
+			}
+		}
+
+		log::info!(target: "UnitsWithAssetType", "almost done");
+
+		// Return the weight used. For each migrated mapping there is a read to get it into
+		// memory, a read to get assetType and
+		// a write to clear the old stored value, and a write to re-store it.
+		let db_weights = T::DbWeight::get();
+		migrated_count.saturating_mul(2 * db_weights.write + 2 * db_weights.read)
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<(), &'static str> {
+		use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+		let pallet_prefix: &[u8] = b"AssetManager";
+		let storage_item_prefix: &[u8] = b"AssetIdUnitsPerSecond";
+
+		// We want to test that:
+		// There are no entries in the new storage beforehand
+		// The same number of mappings exist before and after
+		// As long as there are some mappings stored, one representative key maps to the
+		// same value after the migration.
+		// There are no entries in the old storage afterward
+
+		// Assert new storage is empty
+		assert!(AssetTypeUnitsPerSecond::<T>::iter().next().is_none());
+
+		// Check number of entries, and set it aside in temp storage
+		let stored_data: Vec<_> = storage_key_iter::<T::AssetId, u128, Blake2_128Concat>(
+			pallet_prefix,
+			storage_item_prefix,
+		)
+		.collect();
+		let mapping_count = stored_data.len();
+		Self::set_temp_storage(mapping_count as u32, "mapping_count");
+
+		// Read an example pair from old storage and set it aside in temp storage
+		if mapping_count > 0 {
+			let example_pair = stored_data
+				.iter()
+				.next()
+				.expect("We already confirmed that there was at least one item stored");
+
+			Self::set_temp_storage(example_pair, "example_pair");
+		}
+
+		Ok(())
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade() -> Result<(), &'static str> {
+		use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+		// Check number of entries matches what was set aside in pre_upgrade
+		let old_mapping_count: u64 = Self::get_temp_storage("mapping_count")
+			.expect("We stored a mapping count; it should be there; qed");
+		let new_mapping_count = AssetTypeUnitsPerSecond::<T>::iter().count() as u64;
+		assert_eq!(old_mapping_count, new_mapping_count);
+
+		// Check that our example pair is still well-mapped after the migration
+		if new_mapping_count > 0 {
+			let (asset_id, units): (T::AssetId, u128) =
+				Self::get_temp_storage("example_pair").expect("qed");
+
+			let asset_type =
+				AssetIdType::<T>::get(asset_id).expect("AssetIdType should have the assetType");
+
+			let migrated_units = AssetTypeUnitsPerSecond::<T>::get(asset_type).expect("qed");
+			// Check units are identical
+			assert_eq!(migrated_units, units);
+		}
+
+		Ok(())
+	}
+}
+
+pub struct PopulateAssetTypeIdStorage<T>(PhantomData<T>);
+impl<T: Config> OnRuntimeUpgrade for PopulateAssetTypeIdStorage<T> {
+	fn on_runtime_upgrade() -> Weight {
+		log::info!(target: "PopulateAssetTypeIdStorage", "actually running it");
```

Comment:
I think log statements like this are fine to leave in, but I might change it so that it will have enough context to make sense in the real world